### PR TITLE
dependency: bump Saxon-HE from 9.9.1-7 to 10.2

### PR DESCRIPTION
**Pull request by timurt**

### CODE REVIEW

**Comment by romani** [comment]
There is already 10.2 https://travis-ci.org/github/checkstyle/checkstyle/jobs/723570861#L1130
, does it make sense to try 10.2, or should we do this if few steps ?

**Comment by timurt** [comment]
@romani 
Updated to 10.2, seems ok

**Comment by romani** [comment]
Please update commit message

**Comment by romani** [comment]
commit message is updated.

**Review by romani — APPROVED** [review]
ok to merge.

**Review by pbludov — CHANGES_REQUESTED** [review]
@timurt please try to fix this.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 1]
```diff
@@ -230,7 +230,9 @@ public AxisIterator iterateAxis(byte axisNumber) {
                     }
                 }
                 else {
-                    result = EmptyIterator.OfNodes.THE_INSTANCE;
+                    try (AxisIterator iterator = EmptyIterator.ofNodes()) {
+                        result = iterator;
```
[the remark this answers is not in the document]
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: good point, thanks

rewrote all try-with-resources statements, please check

**Comment by romani** [comment]
CI red state can be ignored.
@pbludov , please finish review

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java [hunk 2]
```diff
@@ -143,9 +143,11 @@ public NodeInfo getRoot() {
      *
      * @param axisNumber element from {@code AxisInfo}
      * @return {@code AxisIterator} object
+     * @noinspection resource, IOResourceOpenedButNotSafelyClosed
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:146]: Please explain reason of suppression in javadoc

**Review by romani — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java (hunk 2 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:146]: @timurt there is a bug in the eclipse compiler related to this issue:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=383965
Please note this in the method Javadoc.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 3]
```diff
@@ -330,7 +310,9 @@ public DetailAST getUnderlyingNode() {
     private AxisIterator getPrecedingSiblingsIterator() {
         final AxisIterator result;
         if (indexAmongSiblings == 0) {
-            result = EmptyIterator.OfNodes.THE_INSTANCE;
+            try (AxisIterator iterator = EmptyIterator.ofNodes()) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java:313]: same here please

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 4]
```diff
@@ -349,7 +331,9 @@ private AxisIterator getPrecedingSiblingsIterator() {
     private AxisIterator getFollowingSiblingsIterator() {
         final AxisIterator result;
         if (indexAmongSiblings == parent.getChildren().size() - 1) {
-            result = EmptyIterator.OfNodes.THE_INSTANCE;
+            try (AxisIterator iterator = EmptyIterator.ofNodes()) {
+                result = iterator;
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java:335]: and here too

**Review by pbludov — CHANGES_REQUESTED** [review]
Two minors

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java (hunk 2 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:150]: @romani done

Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:150]: @pbludov do I really need to specify this bug in javadoc, isn't it CI/CD problem?

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 3 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 4 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java (hunk 2 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:150]: My point is that there is no need for suppression. But due to bugs, some ci tools reports a violation here. One day these bugs will be fixed and the suppression will become redundant. So I suggested mentioning the bug here.

**Review by pbludov — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 5]
```diff
@@ -200,65 +200,51 @@ public NodeInfo getRoot() {
      * Determines axis iteration algorithm. Throws {@code UnsupportedOperationException} in case,
      * when there is no axis iterator for given axisNumber.
      *
+     * Suppresses resource, IOResourceOpenedButNotSafelyClosed: {@link AxisIterator} implements
+     * {@link java.io.Closeable} interface, but none of the subclasses of the {@link AxisIterator}
+     * class has non-empty {@code close()} method.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java:205]: on this method we do not suppress anything, please remove this comment.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 5 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: done, 
I was referring to your previous comment "Please explain reason of suppression in javadoc", how should I explain this IDEA suppressions?

**Comment by romani** [comment]
@timurt , to fix Travis , please rebase on latest master.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java:206]: I pushed a change

**Comment by romani** [comment]
commit message is fixed

**Pull request merged**